x-pack/plugins/fleet/server/services/output.ts

```diff
@@ -104,13 +108,50 @@ async function validateLogstashOutputNotUsedInAPMPolicy(
     perPage: SO_SEARCH_LIMIT,
     withPackagePolicies: true,
   });
-  for (const agentPolicy of agentPolicySO.items) {
-    if (agentPolicyService.hasAPMIntegration(agentPolicy)) {
-      throw new OutputInvalidError('Logstash output cannot be used with APM integration.');
+  return agentPolicySO.items;
+}
+
+async function validateLogstashOutputNotUsedInAPMPolicy(
+  soClient: SavedObjectsClientContract,
+  outputId?: string,
+  isDefault?: boolean
+) {
+  const agentPolicies = await getAgentPoliciesPerOutput(soClient, outputId, isDefault);
+
+  // Validate no policy with APM use that policy
+  if (agentPolicies) {
+    for (const agentPolicy of agentPolicies) {
+      if (agentPolicyService.hasAPMIntegration(agentPolicy)) {
+        throw new OutputInvalidError('Logstash output cannot be used with APM integration.');
+      }
     }
   }
 }
 
+async function findPoliciesWithFleetServer(
+  soClient: SavedObjectsClientContract,
+  outputId?: string,
+  isDefault?: boolean
+) {
+  const agentPolicies = await getAgentPoliciesPerOutput(soClient, outputId, isDefault);
```

Comment:
Question for the rest of @elastic/fleet team (not to address on that PR) here we frequently search for policies with one integration, for that we have to fetch all agent policies and package policies, this is terribly inefficient should we try to denormalize our agent policy document to get that info there and be able to search agent policies with one integration?

Reply:
I'm a little about how much churn we'd generate by substantially reworking our save objects model here. Maybe we could track "agent policies that include this integration" on the `epm-packages` SO type (denormalized model) and add code to our agent policy CRUD to make sure this gets updated?